File: packages/dart/test/protocol/frames_test.dart
import 'package:dhaam_chat/src/protocol/enums.dart';
import 'package:dhaam_chat/src/protocol/envelope.dart';
import 'package:dhaam_chat/src/protocol/errors.dart';
import 'package:dhaam_chat/src/protocol/frames.dart';
import 'package:test/test.dart';

const String _ulid = '01ARZ3NDEKTSV4RRFFQ69G5FAV';

Map<String, Object?> messageJson({int seq = 1, String id = _ulid}) =>
    <String, Object?>{
      'id': id,
      'sessionId': 's1',
      'senderId': 'u1',
      'senderType': 'CUSTOMER',
      'type': 'TEXT',
      'content': 'hello',
      'seq': seq,
      'createdAt': '2026-08-19T12:00:00.000Z',
    };

Map<String, Object?> replayFrame(int seq) => <String, Object?>{
      'v': 1,
      't': 'message.new',
      'id': _ulid,
      'ts': 1700000000000,
      'd': messageJson(seq: seq),
    };

Map<String, Object?> sessionJson() => <String, Object?>{
      'sessionId': 's1',
      'status': 'ASSIGNED',
      'mode': 'HUMAN',
      'participants': <Object?>[
        <String, Object?>{
          'participantId': 'p1',
          'type': 'AGENT',
          'lastReadAt': '2026-08-19T11:00:00.000Z',
        },
      ],
      'createdAt': '2026-08-19T10:00:00.000Z',
    };

void main() {
  group('outbound payloads', () {
    test('connection.hello carries resumeFrom as an INTEGER seq', () {
      // §8.3 calls resumeFrom "the id/seq of the last frame", which reads as
      // permission to send a ULID. D2 says it is the last applied `seq` and
      // the server validates isInteger. §8.3 is simply stale.
      final Map<String, Object?> d = connectionHelloPayload(
        token: 't',
        publishableKey: 'k',
        resumeFrom: 42,
      );
      expect(d['resumeFrom'], isA<int>());
      expect(d['resumeFrom'], equals(42));
    });

    test('connection.hello omits resumeFrom entirely on a fresh connect', () {
      // Absent means "fresh". Sending an explicit null or a 0 would be read as
      // a resume claim, and 0 specifically means "replay everything".
      final Map<String, Object?> d =
          connectionHelloPayload(token: 't', publishableKey: 'k');
      expect(d.containsKey('resumeFrom'), isFalse);
      expect(d['protocolVersion'], equals(kProtocolVersion));
    });

    test('connection.hello omits newSession/subject/topic when unasked', () {
      // Absent, not `false` and not null. The server branches on the key's
      // presence, so a hello that always carried `newSession: false` would
      // differ on the wire from every hello this client sent before the field
      // existed — for a request nobody made.
      final Map<String, Object?> d =
          connectionHelloPayload(token: 't', publishableKey: 'k');
      expect(d.containsKey('newSession'), isFalse);
      expect(d.containsKey('subject'), isFalse);
      expect(d.containsKey('topic'), isFalse);
    });

    test('connection.hello carries newSession with the subject and topic', () {
      final Map<String, Object?> d = connectionHelloPayload(
        token: 't',
        publishableKey: 'k',
        newSession: true,
        subject: 'Refund for order 41',
        topic: 'Billing',
      );
      expect(d['newSession'], isTrue);
      expect(d['subject'], equals('Refund for order 41'));
      expect(d['topic'], equals('Billing'));
    });

    test('connection.hello can carry newSession with no subject or topic', () {
      // "Start a new conversation" with no topic picked is the ordinary case,
      // and it must not become a hello that asks for nothing.
      final Map<String, Object?> d = connectionHelloPayload(
        token: 't',
        publishableKey: 'k',
        newSession: true,
      );
      expect(d['newSession'], isTrue);
      expect(d.containsKey('subject'), isFalse);
      expect(d.containsKey('topic'), isFalse);
    });

    test('message.send always emits type, which the server requires', () {
      // §6.3 shows type as optional (`opts?: { type?: MessageType }`). The
      // server requires it. A client that follows §6.3 literally fails every
      // plain text send.
      final Map<String, Object?> d = messageSendPayload(content: 'hi');
      expect(d['type'], equals('TEXT'));
      expect(d['content'], equals('hi'));
    });

    test('message.send puts an attachment at the top level, not in metadata',
        () {
      final Map<String, Object?> d = messageSendPayload(
        content: '',
        type: MessageType.image,
        attachment: const AttachmentMetadata(
          url: 'https://example.test/a.png',
          fileName: 'a.png',
          mimeType: 'image/png',
          size: 10,
          mediaType: 'IMAGE',
        ),
        metadata: const <String, Object?>{'caption': 'x'},
      );
      expect(d.containsKey('attachment'), isTrue);
      final Map<String, Object?> metadata =
          d['metadata']! as Map<String, Object?>;
      expect(metadata.containsKey('attachment'), isFalse);
    });

    test('typing and heartbeat payloads are empty objects, not null', () {
      // The server requires `d` to be an object for both. Sending null earns
      // VALIDATION_FAILED.
      expect(typingPayload(), isEmpty);
      expect(heartbeatPayload(), isEmpty);
    });
  });

  group('ChatMessage', () {
    test('decodes and exposes seq as the ordering key', () {
      final ChatMessage message =
          ChatMessage.fromJson(messageJson(seq: 7), 'd');
      expect(message.seq, equals(7));
      expect(message.senderType, equals(SenderType.customer));
      expect(message.createdAt.isUtc, isTrue);
    });

    test('requires seq — without it there is no ordering key', () {
      final Map<String, Object?> json = messageJson()..remove('seq');
      expect(
        () => ChatMessage.fromJson(json, 'd'),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('rejects a createdAt with no timezone', () {
      // DateTime.parse would accept this and read it as DEVICE-LOCAL time.
      final Map<String, Object?> json = messageJson()
        ..['createdAt'] = '2026-08-19T12:00:00';
      expect(
        () => ChatMessage.fromJson(json, 'd'),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('rejects epoch millis in createdAt', () {
      // The mirror image of putting ISO-8601 in the envelope ts.
      final Map<String, Object?> json = messageJson()
        ..['createdAt'] = 1700000000000;
      expect(
        () => ChatMessage.fromJson(json, 'd'),
        throwsA(isA<FrameDecodeException>()),
      );
    });
  });

  group('ConnectionAck', () {
    test('decodes the session snapshot, seq anchor and inline replay', () {
      final ConnectionAck ack = ConnectionAck.fromJson(<String, Object?>{
        'protocolVersion': 1,
        'session': sessionJson(),
        'seq': 12,
        'replay': <Object?>[replayFrame(11), replayFrame(12)],
      });

      expect(ack.protocolVersion, equals(1));
      expect(ack.seq, equals(12));
      expect(ack.replay, hasLength(2));
      expect(ack.session.status, equals(ChatStatus.assigned));
      expect(ack.session.participants.single.participantId, equals('p1'));
      expect(ack.session.participants.single.lastReadAt, isNotNull);
    });

    test('treats an absent replay as empty, not as an error', () {
      // Absent covers four situations: a fresh connect, an already-current
      // client, the over-cap path, and a server-side read/projection failure.
      // Only `seq` separates them.
      final ConnectionAck ack = ConnectionAck.fromJson(<String, Object?>{
        'protocolVersion': 1,
        'session': sessionJson(),
        'seq': 40,
      });
      expect(ack.replay, isEmpty);
      expect(ack.seq, equals(40));
    });

    test('validates every replayed frame with the normal decoder', () {
      // A replayed frame is not more trustworthy for being nested inside an
      // ack. §14 applies to it identically.
      expect(
        () => ConnectionAck.fromJson(<String, Object?>{
          'protocolVersion': 1,
          'session': sessionJson(),
          'seq': 12,
          'replay': <Object?>[
            <String, Object?>{
              'v': 1,
              't': 'message.new',
              'id': 'not-a-ulid',
              'ts': 1700000000000,
              'd': messageJson(),
            },
          ],
        }),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('requires the seq anchor', () {
      expect(
        () => ConnectionAck.fromJson(<String, Object?>{
          'protocolVersion': 1,
          'session': sessionJson(),
        }),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('rejects a session status v1 could not model', () {
      final Map<String, Object?> session = sessionJson()
        ..['status'] = 'ESCALATED';
      expect(
        () => ConnectionAck.fromJson(<String, Object?>{
          'protocolVersion': 1,
          'session': session,
          'seq': 1,
        }),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('accepts the two statuses v1 dropped on the floor', () {
      for (final String status in <String>['RESOLVED', 'ON_HOLD']) {
        final Map<String, Object?> session = sessionJson()..['status'] = status;
        final ConnectionAck ack = ConnectionAck.fromJson(<String, Object?>{
          'protocolVersion': 1,
          'session': session,
          'seq': 1,
        });
        expect(ack.session.status.wire, equals(status));
      }
    });
  });

  group('other server payloads', () {
    test('session.closed distinguishes SWITCHED from MANUAL', () {
      final SessionClosed closed = SessionClosed.fromJson(<String, Object?>{
        'sessionId': 's1',
        'closeReason': 'SWITCHED',
      });
      expect(closed.closeReason, equals(CloseReason.switched));
    });

    test('message.read carries an ISO-8601 watermark', () {
      final MessageRead read = MessageRead.fromJson(<String, Object?>{
        'participantId': 'p1',
        'readAt': '2026-08-19T12:00:00Z',
      });
      expect(read.readAt.isUtc, isTrue);
    });

    test('message.delivered watermarks on seq, and carries a display time', () {
      final MessageDelivered delivered =
          MessageDelivered.fromJson(<String, Object?>{
        'participantId': 'p1',
        'deliveredUpToSeq': 7,
        'deliveredAt': '2026-08-19T12:00:00Z',
      });
      expect(delivered.participantId, equals('p1'));
      expect(delivered.deliveredUpToSeq, equals(7));
      expect(delivered.deliveredAt.isUtc, isTrue);
    });

    test('message.delivered accepts an integral double for deliveredUpToSeq',
        () {
      // On Flutter Web every Dart number is a double, so `"deliveredUpToSeq":
      // 7` arrives as 7.0 on exactly one of the three target platforms.
      final MessageDelivered delivered =
          MessageDelivered.fromJson(<String, Object?>{
        'participantId': 'p1',
        'deliveredUpToSeq': 7.0,
        'deliveredAt': '2026-08-19T12:00:00Z',
      });
      expect(delivered.deliveredUpToSeq, equals(7));
    });

    test('message.delivered requires deliveredUpToSeq — the ordering key', () {
      expect(
        () => MessageDelivered.fromJson(<String, Object?>{
          'participantId': 'p1',
          'deliveredAt': '2026-08-19T12:00:00Z',
        }),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('message.delivered refuses a negative watermark', () {
      // Stricter than the TypeScript validator, which checks only isInteger.
      // The server allocates seq from 1; a negative is not a position any
      // honest peer holds, and adopting one would tick nothing forever.
      expect(
        () => MessageDelivered.fromJson(<String, Object?>{
          'participantId': 'p1',
          'deliveredUpToSeq': -1,
          'deliveredAt': '2026-08-19T12:00:00Z',
        }),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('message.delivered rejects a deliveredAt with no timezone', () {
      expect(
        () => MessageDelivered.fromJson(<String, Object?>{
          'participantId': 'p1',
          'deliveredUpToSeq': 7,
          'deliveredAt': '2026-08-19T12:00:00',
        }),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('message.delivered rejects an empty participantId', () {
      expect(
        () => MessageDelivered.fromJson(<String, Object?>{
          'participantId': '',
          'deliveredUpToSeq': 7,
          'deliveredAt': '2026-08-19T12:00:00Z',
        }),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('ticket.linked carries the id, and a url that may be absent', () {
      final TicketLinked linked = TicketLinked.fromJson(<String, Object?>{
        'ticketId': 'tk_1',
      });
      expect(linked.ticketId, equals('tk_1'));
      expect(linked.ticketUrl, isNull);
    });

    test('ticket.linked reads an explicit null ticketUrl as absent', () {
      // openapi/chat-api.yaml's `Ticket` schema — which names the
      // `ticket.linked` frame directly — declares `url` as
      // `type: [string, "null"]`, so null is a value the contract sanctions.
      // Refusing it would drop the whole frame and lose the ticket id too.
      final TicketLinked linked = TicketLinked.fromJson(<String, Object?>{
        'ticketId': 'tk_1',
        'ticketUrl': null,
      });
      expect(linked.ticketUrl, isNull);
    });

    test('ticket.linked requires a non-empty ticketId', () {
      expect(
        () => TicketLinked.fromJson(<String, Object?>{'ticketId': ''}),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('ticket.linked refuses a non-string ticketUrl', () {
      expect(
        () => TicketLinked.fromJson(<String, Object?>{
          'ticketId': 'tk_1',
          'ticketUrl': 42,
        }),
        throwsA(isA<FrameDecodeException>()),
      );
    });

    test('presence.update leaves lastSeen absent while online', () {
      final PresenceEntry entry = PresenceEntry.fromJson(
        <String, Object?>{'participantId': 'p1', 'status': 'ONLINE'},
        'd',
      );
      expect(entry.status, equals(PresenceStatus.online));
      expect(entry.lastSeen, isNull);
    });

    // `agent.joined`/`agent.left` used to be tested here as
    // `{agentId, agentName?}`. That payload is gone — the v2 wire contract
    // replaced it with HandledBy, where displayName is REQUIRED and the old
    // shape is refused rather than coerced. The replacement coverage lives in
    // handled_by_test.dart alongside the rest of the identity contract.
  });
}

File: packages/core/src/protocol/frames.test.ts
import { describe, expect, it } from 'vitest';
import {
  ALL_FRAME_TYPES,
  CLIENT_TO_SERVER_FRAME_TYPES,
  SERVER_PUSH_FRAME_TYPES,
  SERVER_TO_CLIENT_FRAME_TYPES,
} from './frames.js';
import type { AnyFrame } from './frames.js';

describe('frame type catalog — §7.3', () => {
  it('has exactly the 13 client→server frame types', () => {
    expect(CLIENT_TO_SERVER_FRAME_TYPES).toEqual([
      'connection.hello',
      'connection.reauth',
      'session.join',
      'session.leave',
      'session.requestAgent',
      'message.send',
      'message.markRead',
      'message.markDelivered',
      'typing.start',
      'typing.stop',
      'presence.set',
      'presence.query',
      'system.heartbeat',
    ]);
  });

  it('has exactly the 13 plain server push frame types (excludes ack/error)', () => {
    expect(SERVER_PUSH_FRAME_TYPES).toEqual([
      'connection.ack',
      'session.updated',
      'session.closed',
      'agent.joined',
      'agent.left',
      'message.new',
      'typing.start',
      'typing.stop',
      'message.read',
      'message.delivered',
      'presence.update',
      'ticket.linked',
      'system.pong',
    ]);
    expect(SERVER_PUSH_FRAME_TYPES).not.toContain('ack');
    expect(SERVER_PUSH_FRAME_TYPES).not.toContain('error');
  });

  it('server→client set adds exactly ack and error to the push set', () => {
    expect(SERVER_TO_CLIENT_FRAME_TYPES).toEqual([...SERVER_PUSH_FRAME_TYPES, 'ack', 'error']);
  });

  it('has 26 distinct frame type strings total (typing.start/stop shared, not double-counted)', () => {
    const distinct = new Set(ALL_FRAME_TYPES);
    expect(distinct.size).toBe(26);
  });

  it('typing.start and typing.stop are the one shared pair used in both directions (§7.3)', () => {
    expect(CLIENT_TO_SERVER_FRAME_TYPES).toContain('typing.start');
    expect(CLIENT_TO_SERVER_FRAME_TYPES).toContain('typing.stop');
    expect(SERVER_PUSH_FRAME_TYPES).toContain('typing.start');
    expect(SERVER_PUSH_FRAME_TYPES).toContain('typing.stop');
  });
});

// ---------------------------------------------------------------------------
// Exhaustive narrowing over the discriminated union.
//
// This function is the compile-time proof the task requires: every member
// of `AnyFrame` is handled by literal `t`, `frame.d` narrows to the exact
// payload shape for that `t` with NO casts anywhere below, and the
// `default` branch's `const exhaustive: never = frame` line fails `tsc`
// the moment a case is added to the catalog without a matching case here.
// If this file typechecks, exhaustive narrowing holds.
// ---------------------------------------------------------------------------

function describeFrame(frame: AnyFrame): string {
  switch (frame.t) {
    case 'connection.hello':
      return `hello:${frame.d.publishableKey}:${frame.d.protocolVersion}`;
    case 'connection.reauth':
      return `reauth:${frame.d.token}`;
    case 'session.join':
      return `join:${frame.d.sessionId}`;
    case 'session.leave':
      return 'leave';
    case 'session.requestAgent':
      return `requestAgent:${frame.d.reason ?? ''}`;
    case 'message.send':
      return `send:${frame.d.content}:${frame.d.type}`;
    case 'message.markRead':
      return `markRead:${frame.d.upToMessageId ?? ''}`;
    case 'message.markDelivered':
      return `markDelivered:${frame.d.upToSeq}`;
    case 'typing.start':
      return `typingStart:${frame.d.participantId ?? ''}`;
    case 'typing.stop':
      return `typingStop:${frame.d.participantId ?? ''}`;
    case 'presence.set':
      return `presenceSet:${frame.d.status}`;
    case 'presence.query':
      return `presenceQuery:${(frame.d.participantIds ?? []).length}`;
    case 'system.heartbeat':
      return 'heartbeat';
    case 'connection.ack':
      return `ack:${frame.d.seq}:${frame.d.session.sessionId}`;
    case 'session.updated':
      return `sessionUpdated:${frame.d.session.status}`;
    case 'session.closed':
      return `sessionClosed:${frame.d.closeReason}`;
    case 'agent.joined':
      return `agentJoined:${frame.d.kind}:${frame.d.id}:${frame.d.displayName}`;
    case 'agent.left':
      return `agentLeft:${frame.d.kind}:${frame.d.id}:${frame.d.displayName}`;
    case 'message.new':
      return `messageNew:${frame.d.id}:${frame.d.seq}`;
    case 'message.read':
      return `messageRead:${frame.d.participantId}:${frame.d.readAt}`;
    case 'message.delivered':
      return `messageDelivered:${frame.d.participantId}:${frame.d.deliveredUpToSeq}`;
    case 'presence.update':
      return `presenceUpdate:${frame.d.participantId}:${frame.d.status}`;
    case 'ticket.linked':
      return `ticketLinked:${frame.d.ticketId}`;
    case 'system.pong':
      return 'pong';
    case 'ack': {
      if (frame.d.ok) {
        return 'genericAckOk';
      }
      return `genericAckErr:${frame.d.error.code}`;
    }
    case 'error':
      return `error:${frame.d.code}`;
    default: {
      // If this line fails to typecheck, a frame type was added to the
      // catalog (frames.ts) without a matching `case` above.
      const exhaustive: never = frame;
      throw new Error(`unhandled frame type: ${JSON.stringify(exhaustive)}`);
    }
  }
}

describe('exhaustive switch(frame.t) narrowing', () => {
  it('narrows every client→server frame type correctly', () => {
    expect(
      describeFrame({
        v: 1,
        t: 'connection.hello',
        id: '01ARZ3NDEKTSV4RRFFQ69G5FAV',
        ts: 1,
        d: { token: 'tok', publishableKey: 'dhp_test_1', protocolVersion: 1 },
      }),
    ).toBe('hello:dhp_test_1:1');

    expect(
      describeFrame({
        v: 1,
        t: 'message.send',
        id: '01ARZ3NDEKTSV4RRFFQ69G5FAV',
        ts: 1,
        d: { content: 'hi', type: 'TEXT' },
      }),
    ).toBe('send:hi:TEXT');

    expect(
      describeFrame({
        v: 1,
        t: 'session.leave',
        id: '01ARZ3NDEKTSV4RRFFQ69G5FAV',
        ts: 1,
        d: {},
      }),
    ).toBe('leave');
  });

  it('narrows every server→client push frame type correctly', () => {
    expect(
      describeFrame({
        v: 1,
        t: 'connection.ack',
        id: '01ARZ3NDEKTSV4RRFFQ69G5FAV',
        ts: 1,
        d: {
          protocolVersion: 1,
          seq: 42,
          session: {
            sessionId: 'sess_1',
            status: 'OPEN',
            mode: 'BOT',
            participants: [],
            createdAt: '2026-08-17T00:00:00.000Z',
          },
        },
      }),
    ).toBe('ack:42:sess_1');

    expect(
      describeFrame({
        v: 1,
        t: 'session.closed',
        id: '01ARZ3NDEKTSV4RRFFQ69G5FAV',
        ts: 1,
        d: { sessionId: 'sess_1', closeReason: 'SWITCHED' },
      }),
    ).toBe('sessionClosed:SWITCHED');
  });

  it('narrows the generic ack frame on both the ok:true and ok:false branches', () => {
    expect(
      describeFrame({
        v: 1,
        t: 'ack',
        id: '01ARZ3NDEKTSV4RRFFQ69G5FAW',
        ref: '01ARZ3NDEKTSV4RRFFQ69G5FAV',
        ts: 1,
        d: { ok: true, seq: 7 },
      }),
    ).toBe('genericAckOk');

    expect(
      describeFrame({
        v: 1,
        t: 'ack',
        id: '01ARZ3NDEKTSV4RRFFQ69G5FAW',
        ref: '01ARZ3NDEKTSV4RRFFQ69G5FAV',
        ts: 1,
        d: { ok: false, error: { code: 'SESSION_CLOSED', message: 'closed', retryable: false } },
      }),
    ).toBe('genericAckErr:SESSION_CLOSED');
  });

  it('narrows the error frame', () => {
    expect(
      describeFrame({
        v: 1,
        t: 'error',
        id: '01ARZ3NDEKTSV4RRFFQ69G5FAV',
        ts: 1,
        d: { code: 'AUTH_EXPIRED', message: 'expired', retryable: true },
      }),
    ).toBe('error:AUTH_EXPIRED');
  });
});
